File: intellij-plugin-structure/structure-classes/src/main/java/com/jetbrains/plugin/structure/impl/resolvers/JdkResolverCreator.java
package com.jetbrains.plugin.structure.impl.resolvers;

import com.google.common.base.Predicate;
import com.google.common.collect.ImmutableSet;
import com.jetbrains.plugin.structure.impl.utils.JarsUtils;
import com.jetbrains.plugin.structure.resolvers.Resolver;
import org.jetbrains.annotations.NotNull;

import java.io.File;
import java.io.IOException;
import java.util.Collection;
import java.util.Set;

/**
 * @author Sergey Patrikeev
 */
public class JdkResolverCreator {
  private static final Set<String> JDK_JAR_NAMES = ImmutableSet.of("rt.jar", "tools.jar", "classes.jar", "jsse.jar", "javaws.jar", "jce.jar", "jfxrt.jar", "plugin.jar");

  @NotNull
  public static Resolver createJdkResolver(@NotNull File jdkDir) throws IOException {
    Collection<File> jars = JarsUtils.collectJars(jdkDir, new Predicate<File>() {
      @Override
      public boolean apply(File file) {
        return JDK_JAR_NAMES.contains(file.getName().toLowerCase());
      }
    }, true);

    return JarsUtils.makeResolver("Jdk resolver " + jdkDir.getCanonicalPath(), jars);
  }

}

File: intellij-plugin-structure/structure-classes/src/main/java/com/jetbrains/plugin/structure/classes/jdk/JdkResolverCreator.kt
package com.jetbrains.plugin.structure.classes.jdk

import com.jetbrains.plugin.structure.classes.resolvers.Resolver
import com.jetbrains.plugin.structure.classes.utils.JarsUtils

import java.io.File

object JdkResolverCreator {

  private val MANDATORY_JARS = setOf("rt.jar")

  private val ADDITIONAL_JARS = setOf("tools.jar", "classes.jar", "jsse.jar", "javaws.jar", "jce.jar", "jfxrt.jar", "plugin.jar")

  fun createJdkResolver(jdkPath: File): Resolver = createJdkResolver(Resolver.ReadMode.FULL, jdkPath)

  fun createJdkResolver(readMode: Resolver.ReadMode, jdkPath: File): Resolver {
    val mandatoryJars = JarsUtils.collectJars(jdkPath, { it.name.toLowerCase() in MANDATORY_JARS }, true)
    val missingJars = MANDATORY_JARS - mandatoryJars.map { it.name }
    if (missingJars.isNotEmpty()) {
      throw IllegalArgumentException("JDK ${jdkPath.absolutePath} is not a valid JDK")
    }

    val additionalJars = JarsUtils.collectJars(jdkPath, { it.name.toLowerCase() in ADDITIONAL_JARS }, true)
    return JarsUtils.makeResolver(readMode, mandatoryJars + additionalJars)
  }

}
